src/diff.rs

```diff
@@ -1,103 +1,252 @@
+                        move_element(actual_elem, &mut missing, &mut extra);
+                    }
+                    (None, Some(expect_elem)) => {
+                        move_element(expect_elem, &mut extra, &mut missing);
+                    }
+                    (Some(actual_elem), None) => {
+                        move_element(actual_elem, &mut missing, &mut extra);
+                    }
+                    (None, None) => break,
+                }
+            }
+            SequenceComparison {
+                order_preserved,
+                exclusive_left: extra,
+                exclusive_right: missing,
+            }
+        }
+
+        pub(self) fn relative_order_comparison<
+            ICL: Iterator<Item = T> + Clone,
+            ICR: Iterator<Item = T> + Clone,
+        >(
+            mut actual_iter: ICL,
+            mut expected_iter: ICR,
+        ) -> SequenceComparison<T> {
+            let mut actual_value = actual_iter.next();
+            let mut expected_value = expected_iter.next();
+            let mut missing: Vec<T> = vec![];
+            let mut extra: Vec<T> = vec![];
+            loop {
+                if expected_value.is_none() {
+                    extra.extend(actual_iter);
+                    break;
+                }
+                if actual_value.is_none() {
+                    missing.push(expected_value.unwrap());
+                    missing.extend(expected_iter);
+                    break;
+                }
+                if actual_value.eq(&expected_value) {
+                    actual_value = actual_iter.next();
+                    expected_value = expected_iter.next();
+                } else {
+                    extra.push(actual_value.unwrap());
+                    actual_value = actual_iter.next();
+                }
+            }
+            let order_preserved = missing.is_empty();
+
+            // check out of order elements.
+            if !missing.is_empty() {
+                for extra_elem in extra.iter() {
+                    if let Some(idx) = missing.iter().position(|m: &T| m.eq(extra_elem)) {
+                        missing.remove(idx);
+                    }
+                }
+            }
+
+            SequenceComparison {
+                order_preserved,
+                exclusive_left: extra,
+                exclusive_right: missing,
+            }
+        }
     }
 
-    #[test]
-    fn map_diff_common() {
-        let left: HashMap<&str, i32> = HashMap::from([("123", 2)]);
-        let right: HashMap<&str, i32> = HashMap::from([("123", 2)]);
-        let result = MapComparison::from_hash_maps(&left, &right);
-        assert_eq!(result.common, vec![(&"123", &2)]);
-        assert!(result.exclusive_left.is_empty());
-        assert!(result.exclusive_right.is_empty());
+    #[cfg(test)]
+    mod tests {
+        // TODO: add quickcheck and parameterized (test_case / rstest) tests; for now covered with public API tests
```

Comment:
Could you write test?  It is ok after discussion for impl ends.

Reply:
I wanted to do this in separate PR to keep the size managable, do you want me to add parameterized tests in this one?

Reply:
Could you add tests in one PR to catch such bugs we've found?  As this is assertion library, the reliability of code is important.